osutil/epoll/epoll_test.go

```diff
@@ -0,0 +1,546 @@
+package epoll_test
+
+import (
+	"os"
+	"syscall"
+	"testing"
+	"time"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/snapcore/snapd/osutil/epoll"
+
+	. "gopkg.in/check.v1"
+)
+
+func Test(t *testing.T) { TestingT(t) }
+
+type epollSuite struct{}
+
+var _ = Suite(&epollSuite{})
+
+func (*epollSuite) TestString(c *C) {
+	c.Check(epoll.Readable.String(), Equals, "Readable")
+	c.Check(epoll.Writable.String(), Equals, "Writable")
+	c.Check(epoll.Readiness(epoll.Readable|epoll.Writable).String(), Equals, "Readable|Writable")
+}
+
+func (*epollSuite) TestOpenClose(c *C) {
+	e, err := epoll.Open()
+	c.Assert(err, IsNil)
+	c.Assert(e.Fd() == -1, Equals, false)
+	c.Assert(e.Fd() == 0, Equals, false)
+	c.Assert(e.Fd() == 1, Equals, false)
+	c.Assert(e.Fd() == 2, Equals, false)
+	c.Assert(e.RegisteredFdCount(), Equals, 0)
+
+	err = e.Close()
+	c.Assert(err, IsNil)
+	c.Assert(e.Fd(), Equals, -1)
+}
+
+func concurrentlyRegister(e *epoll.Epoll, fd int, errCh chan error) {
+	err := e.Register(fd, epoll.Readable)
+	errCh <- err
+}
+
+func concurrentlyDeregister(e *epoll.Epoll, fd int, errCh chan error) {
+	err := e.Deregister(fd)
+	errCh <- err
+}
+
+func waitMillisecondsThenWriteToFd(msec int, fd int, msg []byte) error {
+	time.Sleep(time.Duration(msec) * time.Millisecond)
+	_, err := unix.Write(fd, msg)
+	return err
+}
+
+func (*epollSuite) TestRegisterWaitModifyDeregister(c *C) {
+	e, err := epoll.Open()
+	c.Assert(err, IsNil)
+
+	socketFds, err := unix.Socketpair(unix.AF_UNIX, unix.SOCK_STREAM, 0)
+	c.Assert(err, IsNil)
+	defer unix.Close(socketFds[0])
+	defer unix.Close(socketFds[1])
+
+	listenerFd := socketFds[0]
+	senderFd := socketFds[1]
+
+	err = unix.SetNonblock(listenerFd, true)
+	c.Assert(err, IsNil)
+
+	err = e.Register(listenerFd, epoll.Readable)
+	c.Assert(err, IsNil)
+
+	msg := []byte("foo")
+
+	go waitMillisecondsThenWriteToFd(1000, senderFd, msg)
```

Comment:
(super nitpick) Maybe slightly shorter timeouts? 250ms? The total runtime is currently 6.7s of the testsuite - maybe we could use `var defaultWait = testutil.HostScaledTimeout(200 * time.Millisecond)` everywhere where we currently use `1000 * time.Millisecond` ?